From 08b93f42b4eac65a8c9580e5e9ece522328fde15 Mon Sep 17 00:00:00 2001
From: Ryan Gonzalez <rymg19@gmail.com>
Date: Wed, 11 Nov 2020 16:54:18 -0600
Subject: [PATCH 17/21] [webusb] Properly handle lack of platform support

This avoids crashes and infinite device service startup loops when the
platform does not have a USB service, which can occur on builds without
udev.
---
 chrome/browser/usb/web_usb_detector.cc        |  1 +
 .../device/usb/mojo/device_manager_impl.cc    | 27 +++++++++++++------
 2 files changed, 20 insertions(+), 8 deletions(-)

diff --git a/chrome/browser/usb/web_usb_detector.cc b/chrome/browser/usb/web_usb_detector.cc
index 4bc06f664534c..0c8844b73985c 100644
--- a/chrome/browser/usb/web_usb_detector.cc
+++ b/chrome/browser/usb/web_usb_detector.cc
@@ -275,6 +275,7 @@ void WebUsbDetector::OnDeviceManagerConnectionError() {
   client_receiver_.reset();
 
   // Try to reconnect the device manager.
+  LOG(INFO) << "Restarting USB device manager after connection error";
   Initialize();
 }
 
diff --git a/services/device/usb/mojo/device_manager_impl.cc b/services/device/usb/mojo/device_manager_impl.cc
index f388efa43223d..02b382c6923c0 100644
--- a/services/device/usb/mojo/device_manager_impl.cc
+++ b/services/device/usb/mojo/device_manager_impl.cc
@@ -45,29 +45,40 @@ DeviceManagerImpl::~DeviceManagerImpl() = default;
 
 void DeviceManagerImpl::AddReceiver(
     mojo::PendingReceiver<mojom::UsbDeviceManager> receiver) {
-  if (usb_service_)
-    receivers_.Add(this, std::move(receiver));
+  receivers_.Add(this, std::move(receiver));
 }
 
 void DeviceManagerImpl::EnumerateDevicesAndSetClient(
     mojo::PendingAssociatedRemote<mojom::UsbDeviceManagerClient> client,
     EnumerateDevicesAndSetClientCallback callback) {
-  usb_service_->GetDevices(base::BindOnce(
-      &DeviceManagerImpl::OnGetDevices, weak_factory_.GetWeakPtr(),
-      /*options=*/nullptr, std::move(client), std::move(callback)));
+  if (usb_service_) {
+    usb_service_->GetDevices(base::BindOnce(
+        &DeviceManagerImpl::OnGetDevices, weak_factory_.GetWeakPtr(),
+        /*options=*/nullptr, std::move(client), std::move(callback)));
+  } else {
+    OnGetDevices(nullptr, std::move(client), std::move(callback), {});
+  }
 }
 
 void DeviceManagerImpl::GetDevices(mojom::UsbEnumerationOptionsPtr options,
                                    GetDevicesCallback callback) {
-  usb_service_->GetDevices(base::BindOnce(
-      &DeviceManagerImpl::OnGetDevices, weak_factory_.GetWeakPtr(),
-      std::move(options), mojo::NullAssociatedRemote(), std::move(callback)));
+  if (usb_service_) {
+    usb_service_->GetDevices(base::BindOnce(
+        &DeviceManagerImpl::OnGetDevices, weak_factory_.GetWeakPtr(),
+        std::move(options), mojo::NullAssociatedRemote(), std::move(callback)));
+  } else {
+    OnGetDevices(std::move(options), mojo::NullAssociatedRemote(),
+                 std::move(callback), {});
+  }
 }
 
 void DeviceManagerImpl::GetDevice(
     const std::string& guid,
     mojo::PendingReceiver<mojom::UsbDevice> device_receiver,
     mojo::PendingRemote<mojom::UsbDeviceClient> device_client) {
+  if (!usb_service_)
+    return;
+
   scoped_refptr<UsbDevice> device = usb_service_->GetDevice(guid);
   if (!device)
     return;
-- 
2.26.2

